### PR TITLE
Adding Serialization

### PR DESCRIPTION
**Pull request by djw8605**

### CODE REVIEW

File: src/scitokens/scitokens.py [hunk 1]
```diff
@@ -30,17 +37,22 @@ def decode_base64(data):
...
 class UnsupportedKeyException(Exception):
     pass
+    
+class MissingIssuerException(Exception):
+    pass
 
 class SciToken(object):
 
-    def __init__(self, key=None, parent=None):
+    def __init__(self, key=None, parent=None, claims=None, serialized_token=None):
```
Comment (bbockelm) [inline src/scitokens/scitokens.py]: Drop this for now -- wait until we decide how to use it.

File: src/scitokens/scitokens.py [hunk 2]
```diff
@@ -74,13 +89,42 @@ def verify(self):
         raise NotImplementedError()
 
 
-    def serialize(self, include_key=False):
+    def serialize(self, include_key=False, issuer = None, lifetime = 600):
```
Comment (bbockelm) [inline src/scitokens/scitokens.py]: Remove extra spaces.

File: src/scitokens/scitokens.py [hunk 3]
```diff
@@ -74,13 +89,42 @@ def verify(self):
...
+            raise MissingKeyException("Unable to serialize, missing private key")
+        
+        # Issuer needs to be available, otherwise throw an error
+        if issuer == None and 'iss' not in self._claims:
+            raise MissingIssuerException("Issuer not specific in claims or as argument")
+        
+        if not issuer:
+            issuer = self._claims['iss']
+        
+        issue_time = int(mktime(datetime.utcnow().timetuple()))
```
Comment (bbockelm) [inline src/scitokens/scitokens.py]: Allow these claims to be set externally - only set them if they are not already present.

Further, change to `time.time()`?

File: src/scitokens/scitokens.py [hunk 4]
```diff
@@ -74,13 +89,42 @@ def verify(self):
...
+        payload = dict(self._claims)
+        
+        # Anything below will override what is in the claims
+        payload.update({
+            "iss": issuer,
+            "exp": exp_time,
+            "iat": issue_time,
+        })
+        
+        encoded = jwt.encode(payload, self._key, algorithm = "RS256")
```
Comment (bbockelm) [inline src/scitokens/scitokens.py]: Once you've signed the token, save it to a `_serialized_token` member.  Further, copy over all the claims to the verified claims dict.

File: src/scitokens/scitokens.py [hunk 5]
```diff
@@ -155,22 +212,29 @@ def deserialize(serialized_token, require_key=True):
...
         
         # Get the public key from the issuer
         issuer_public_key = SciToken._get_issuer_publickey(unverified_headers, unverified_payload)
         
         claims = jwt.decode(serialized_token, issuer_public_key)
-        print(claims)
-        return
+        # Do we have the private key?
+        if (len(info) == 4):
+            to_return = SciToken(serialized_token = serialized_token, key = key)
```
Comment (bbockelm) [inline src/scitokens/scitokens.py]: Create the token, then set the member variable directly.  Let's not add `serialized_token` to the constructor API.

File: src/scitokens/scitokens.py [hunk 6]
```diff
@@ -155,22 +212,29 @@ def deserialize(serialized_token, require_key=True):
...
+        else:
+            to_return = SciToken(serialized_token = serialized_token)
+            
+        to_return.verified_claims = claims
```
Comment (bbockelm) [inline src/scitokens/scitokens.py]: Use `_verified_claims`

File: tests/test_create_scitoken.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        ))
+        
+        token = scitokens.SciToken(key = private_key)
+        token.update_claims({"test": "true"})
+        serialized_token = token.serialize(issuer = "local")
+        
+        self.assertEqual(len(serialized_token.decode('utf8').split(".")), 3)
+        print(serialized_token)
+        
+
```
Comment (bbockelm) [inline tests/test_create_scitoken.py]: Remove excess whitespace.  Use 2 blank lines.

File: tests/test_with_http.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+# For use in the HTTP Serve test class
+#test_kid = ""
+test_n = 0
+test_e = 0
+
+def bytes_from_long(data):
+    return base64.urlsafe_b64encode(cryptography.utils.int_to_bytes(data)).decode('ascii')
+
+class S(BaseHTTPRequestHandler):
```
Comment (bbockelm) [inline tests/test_with_http.py]: Better class name?

File: tests/test_with_http.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+class TestDeserialization(unittest.TestCase):
+
+    def setUp(self):
+        # Start a web server to act as the "issuer"
+        server_address = ('', 8080)
+        httpd = HTTPServer(server_address, S)
+        t = threading.Thread(target = httpd.serve_forever)
+        t.daemon = True
+        t.start()
```
Comment (bbockelm) [inline tests/test_with_http.py]: Destroy object in `tearDown` so we get a fresh server each test.

File: src/scitokens/scitokens.py (hunk 4 above)
Comment (djw8605) [reply src/scitokens/scitokens.py]: Done!

**Comment by bbockelm** [comment]
@djw8605 - some merge conflicts are noted by github UI.

**Comment by bbockelm** [comment]
Per our discussion over IM, I am going to override the Codacy check just so we can move things forward.

**Pull request merged**